# 1. Simple Route to Greet the User

In [ ]:
# !pip install flask

## Explanation of the Flask Code Snippet

This is a minimal Flask web application that creates a simple HTTP server. Let me break it down:

### 1. Import Statement
```python
from flask import Flask
```
Imports the `Flask` class from the flask library. This is the core class needed to create a web application.

### 2. Application Instance
```python
app = Flask(__name__)
```
Creates a Flask application instance. The `__name__` argument tells Flask where to look for resources like templates and static files. When run directly, `__name__` equals `'__main__'`.

### 3. Route Decorator
```python
@app.route('/')
def greet():
    return 'Hello, World!'
```
- **`@app.route('/')`** - A decorator that binds the URL path `/` (the root URL) to the `greet()` function
- **`def greet()`** - The view function that handles requests to this route
- **`return 'Hello, World!'`** - Returns a plain text response to the browser

### 4. Entry Point
```python
if __name__ == '__main__':
    app.run(debug=False)
```
- This conditional ensures the server only runs when the script is executed directly (not when imported as a module)
- `app.run()` starts Flask's built-in development server
- `debug=False` disables debug mode (in production, you'd use a WSGI server like Gunicorn instead)

### 5. Testing the Endpoint
```bash
curl http://localhost:5000/
```
This curl command sends a GET request to the running server. By default, Flask runs on port **5000** on localhost.

---

**Flow Summary:**  
User visits `http://localhost:5000/` → Flask matches the `/` route → Calls `greet()` → Returns `"Hello, World!"` to the browser.

In [1]:
from flask import Flask

app = Flask(__name__)

@app.route('/')
def greet():
    return 'Hello, World!'

if __name__ == '__main__':
    app.run(debug=False)


### Accessing the endpoint via curl (Commandline / Terminal)
# curl http://localhost:5000/

 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [17/Dec/2025 02:42:59] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [17/Dec/2025 02:43:00] "GET /favicon.ico HTTP/1.1" 404 -


# 2. Route with User Input and Computation
- Let's create a route that takes a number as input, squares it, and displays the result.
- Also `/greet/<name>` route to your Flask application, which takes the username from the URL and greets them.

- We use `render_template_string` to render an HTML template with the result.
- `{{variable}}` syntax is from `Jinja2`, used for templating.

## Explanation of the Flask Code with Dynamic Routes

This Flask application demonstrates **dynamic URL routing** - capturing values from the URL and using them in your functions.

### 1. Imports
```python
from flask import Flask, request, render_template_string
```
- **`Flask`** - Core class to create the app
- **`request`** - Access incoming request data (not used in this snippet but imported)
- **`render_template_string`** - Renders HTML templates directly from strings using Jinja2

### 2. Application Instance
```python
app = Flask(__name__)
```
Creates the Flask application instance.

---

### 3. Dynamic Route with String Parameter
```python
@app.route('/greet/<name>')
def greet_user(name):
    return render_template_string('<h1>Hello, {{name}}!</h1>', name=name)
```
- **`<name>`** - A **variable rule** that captures any string from the URL
- The captured value is passed as an argument to `greet_user(name)`
- **`{{name}}`** - Jinja2 template syntax to insert the variable into HTML

**Example:** `/greet/Prashant` → `Hello, Prashant!`

---

### 4. Dynamic Route with Integer Parameter
```python
@app.route('/square/<int:number>')
def square(number):
    result = number ** 2
    return render_template_string('<h1>{{number}} squared is {{result}}</h1>', number=number, result=result)
```
- **`<int:number>`** - A **converter** that:
  - Captures only integers from the URL
  - Automatically converts the string to an `int`
  - Returns 404 if the value isn't a valid integer
- Computes `number ** 2` and displays the result

**Example:** `/square/5` → `5 squared is 25`

---

### 5. Entry Point & Testing
```python
if __name__ == '__main__':
    app.run(debug=False)
```

**Test with curl:**
```bash
curl http://localhost:5000/greet/Prashant    # Returns: <h1>Hello, Prashant!</h1>
curl http://localhost:5000/square/5          # Returns: <h1>5 squared is 25</h1>
```

---

### URL Converters Available in Flask

| Converter | Description |
|-----------|-------------|
| `string` | (default) Accepts any text without slashes |
| `int` | Accepts positive integers |
| `float` | Accepts positive floating point values |
| `path` | Like string but also accepts slashes |
| `uuid` | Accepts UUID strings |

In [ ]:
from flask import Flask, request, render_template_string

app = Flask(__name__)

# Captures a string name from the URL.
@app.route('/greet/<name>')
def greet_user(name):
    return render_template_string('<h1>Hello, {{name}}!</h1>', name=name)

# The route /square/<int:number> captures an integer from the URL.
@app.route('/square/<int:number>')
def square(number):
    result = number ** 2
    return render_template_string('<h1>{{number}} squared is {{result}}</h1>', number=number, result=result)

if __name__ == '__main__':
    app.run(debug=False)

    
# Testing with curl
# Replace <name> with any username you like.
# curl http://localhost:5000/greet/Prashant
               
# curl http://localhost:5000/square/5    

 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [17/Dec/2025 02:46:53] "GET / HTTP/1.1" 404 -
127.0.0.1 - - [17/Dec/2025 02:47:05] "GET /greet/sourav HTTP/1.1" 200 -
127.0.0.1 - - [17/Dec/2025 02:47:14] "GET /square/5 HTTP/1.1" 200 -


**Additional Notes:**

- **URL Encoding:**
  - If the name contains spaces or special characters, ensure it's URL-encoded.
  - Refer to: https://www.w3schools.com/tags/ref_urlencode.asp
  - Online URL Encoder: https://www.urlencoder.io/
  - For example, for "John Doe":

    ```bash
    curl http://localhost:5000/greet/John%20Doe
    ```

- **Error Handling:**
  - Currently, the code does not handle errors like missing parameters or invalid inputs.
  - You might want to add error handling to make your application more robust.

**Feel free to ask if you need further clarification or assistance!**

# 3. Route Demonstrating the GET Method
Let's create a route that accepts query parameters using the GET method.

- `400` is the HTTP status code for `Bad Request`.
- By returning `400`, the server indicates to the client that the request was invalid due to missing required parameters.

## Explanation of Flask GET Route with Query Parameters

This Flask application demonstrates how to handle **query parameters** with proper **error handling** and **input validation**.

---

### 1. Imports
```python
from flask import Flask, request, render_template_string
import math
```
- **`request`** - Access incoming request data (query params, form data, etc.)
- **`math`** - Python's math library for the `sqrt()` function

---

### 2. Route Definition
```python
@app.route('/sqrt', methods=['GET'])
def compute_sqrt():
```
- **`methods=['GET']`** - Explicitly specifies this route only accepts GET requests
- Query parameters are passed in the URL after `?` (e.g., `/sqrt?num1=16&num2=4`)

---

### 3. Retrieving Query Parameters
```python
num1 = request.args.get('num1', type=float)
num2 = request.args.get('num2', type=float)
```
- **`request.args`** - A dictionary-like object containing all query parameters
- **`.get('num1', type=float)`** - Retrieves the parameter and automatically converts it to `float`
- Returns `None` if the parameter is missing or can't be converted

---

### 4. Validation & Error Handling

| Check | HTTP Status | Purpose |
|-------|-------------|---------|
| `num1 is None or num2 is None` | **400** | Missing required parameters |
| `num2 == 0` | **400** | Prevent division by zero |
| `division_result < 0` | **400** | Can't compute square root of negative |
| `except ValueError` | **400** | Invalid non-numeric input |

**HTTP 400 = Bad Request** - Tells the client their request was invalid.

---

### 5. Computation Logic
```python
division_result = num1 / num2
sqrt_result = math.sqrt(division_result)
```
First divides `num1` by `num2`, then computes the square root of the result.

---

### 6. Returning Response with Status Code
```python
return render_template_string('<h1>Error message</h1>'), 400
```
The second value (`400`) is the **HTTP status code** returned with the response.

---

### Testing Examples

| curl Command | Result |
|--------------|--------|
| `curl "http://localhost:5000/sqrt?num1=16&num2=4"` | ✅ `Square root of 4.0 is 2.0` |
| `curl "http://localhost:5000/sqrt?num2=4"` | ❌ Missing `num1` parameter |
| `curl "http://localhost:5000/sqrt?num1=16&num2=0"` | ❌ Division by zero |
| `curl "http://localhost:5000/sqrt?num1=-16&num2=4"` | ❌ Negative result (-4) |
| `curl "http://localhost:5000/sqrt?num1=abc&num2=4"` | ❌ Invalid input |

---

### Query Parameters vs URL Parameters

| Feature | Query Parameters | URL Parameters |
|---------|------------------|----------------|
| **Syntax** | `/sqrt?num1=16&num2=4` | `/square/5` |
| **Access** | `request.args.get()` | Function argument |
| **Use case** | Optional filters, search | Required resource identifiers |
| **Visibility** | After `?` in URL | Part of URL path |

In [2]:
from flask import Flask, request, render_template_string
import math

app = Flask(__name__)

@app.route('/sqrt', methods=['GET'])
def compute_sqrt():
    try:
        # Retrieve numbers from query parameters and convert to float
        num1 = request.args.get('num1', type=float)
        num2 = request.args.get('num2', type=float)

        # Check if both numbers are provided
        if num1 is None or num2 is None:
            # Return error message with HTTP 400 status code
            return render_template_string(
                '<h1>Error: Both num1 and num2 parameters are required.</h1>'
            ), 400

        # Prevent division by zero
        if num2 == 0:
            return render_template_string(
                '<h1>Error: Division by zero is not allowed.</h1>'
            ), 400

        # Compute the division of num1 by num2
        division_result = num1 / num2

        # Ensure the result is non-negative for square root
        if division_result < 0:
            return render_template_string(
                '<h1>Error: Cannot compute square root of a negative number.</h1>'
            ), 400

        # Compute the square root
        sqrt_result = math.sqrt(division_result)

        # Return the result
        return render_template_string(
            '<h1>Square root of {{division_result}} is {{sqrt_result}}</h1>',
            division_result=division_result,
            sqrt_result=sqrt_result
        )

    except ValueError:
        # Handle cases where inputs are not numbers
        return render_template_string(
            '<h1>Error: Invalid input. Please provide numeric values for num1 and num2.</h1>'
        ), 400

   
if __name__ == '__main__':
    app.run(debug=False)
    
# Testing with curl
# curl "http://localhost:5000/sqrt?num1=16&num2=4"
# curl "http://localhost:5000/sqrt?num2=4"
# curl "http://localhost:5000/sqrt?num1=16&num2=0"
# curl "http://localhost:5000/sqrt?num1=-16&num2=4"
# curl "http://localhost:5000/sqrt?num1=abc&num2=4"



 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [17/Dec/2025 02:50:29] "GET / HTTP/1.1" 404 -
127.0.0.1 - - [17/Dec/2025 02:50:45] "GET /sqrt?num1=16&num2=4 HTTP/1.1" 200 -


### Accessing via Web Browser:
You can also test the endpoint by navigating to:

- Valid Input: http://localhost:5000/sqrt?num1=16&num2=4
- Missing Parameters: http://localhost:5000/sqrt?num1=16
- Division by Zero: http://localhost:5000/sqrt?num1=16&num2=0
- Negative Result: http://localhost:5000/sqrt?num1=-16&num2=4

# 4. Route Demonstrating the POST Method
We'll create a route that accepts data via POST and displays it.

## Explanation of Flask POST Method with Form Handling

This Flask application demonstrates how to handle **both GET and POST requests** on the same route, implementing a simple form submission flow.

---

### 1. Route Definition
```python
@app.route('/submit', methods=['GET', 'POST'])
def submit():
```
- **`methods=['GET', 'POST']`** - This route accepts both HTTP methods
- Same URL `/submit` behaves differently based on the request method

---

### 2. Handling POST Requests (Form Submission)
```python
if request.method == 'POST':
    message = request.form.get('message', 'No message')
    return render_template_string('<h1>You submitted:<hr /> {{message}}</h1>', message=message)
```
- **`request.method`** - Check which HTTP method was used
- **`request.form`** - A dictionary containing form data sent via POST
- **`.get('message', 'No message')`** - Retrieves the `message` field, defaults to `'No message'` if missing
- **`<hr />`** - HTML horizontal rule for visual separation

---

### 3. Handling GET Requests (Display Form)
```python
else:
    return '''
        <form method="post" action="/submit">
            <input type="text" name="message" placeholder="Enter a message" />
            <input type="submit" />
        </form>
    '''
```
Returns an HTML form with:
- **`method="post"`** - Form data sent via POST
- **`action="/submit"`** - Form submits to the same route
- **`name="message"`** - Input field name (matches `request.form.get('message')`)

---

### Flow Diagram

```
┌─────────────────────────────────────────────────────────────┐
│  User visits /submit (GET)                                  │
│         ↓                                                   │
│  Server returns HTML form                                   │
│         ↓                                                   │
│  User fills form and clicks Submit                          │
│         ↓                                                   │
│  Browser sends POST to /submit with form data               │
│         ↓                                                   │
│  Server extracts message, returns "You submitted: ..."      │
└─────────────────────────────────────────────────────────────┘
```

---

### Testing

**Via Browser:**
1. Navigate to `http://127.0.0.1:5000/submit`
2. Enter a message in the text box
3. Click Submit

**Via curl:**
```bash
# POST with form data
curl -X POST -d "message=Hello Flask" http://localhost:5000/submit
```

---

### GET vs POST Data Access

| Data Source | Access Method | Use Case |
|-------------|---------------|----------|
| Query params (`?key=value`) | `request.args.get()` | GET requests, filters |
| Form data | `request.form.get()` | POST from HTML forms |
| JSON body | `request.get_json()` | API requests |

In [3]:
from flask import Flask, request, render_template_string

app = Flask(__name__)

@app.route('/submit', methods=['GET', 'POST'])
def submit():
    if request.method == 'POST':
        message = request.form.get('message', 'No message')
        return render_template_string('<h1>You submitted:<hr /> {{message}}</h1>', message=message)
    else:
        # Display a simple form
        return '''
            <form method="post" action="/submit">
                <input type="text" name="message" placeholder="Enter a message" />
                <input type="submit" />
            </form>
        '''

if __name__ == '__main__':
    app.run(debug=False)

    
# Access it here: http://127.0.0.1:5000/submit

# curl -X POST -d "your message" http://localhost:5000/submit

 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [17/Dec/2025 02:52:34] "GET / HTTP/1.1" 404 -
127.0.0.1 - - [17/Dec/2025 02:52:42] "GET /submit HTTP/1.1" 200 -
127.0.0.1 - - [17/Dec/2025 02:52:45] "POST /submit HTTP/1.1" 200 -
127.0.0.1 - - [17/Dec/2025 02:53:22] "GET /submit/4 HTTP/1.1" 404 -
127.0.0.1 - - [17/Dec/2025 02:54:04] "GET /submit?4 HTTP/1.1" 200 -
127.0.0.1 - - [17/Dec/2025 02:54:09] "POST /submit HTTP/1.1" 200 -
127.0.0.1 - - [17/Dec/2025 02:56:04] "GET /submit/extra HTTP/1.1" 404 -


In [ ]:
# import requests

# # URL of the Flask endpoint
# url = 'http://127.0.0.1:5000/submit'

# # Data to be sent in the POST request
# data = {
#     'message': 'Hello, Flask!'
# }

# # Make the POST request
# response = requests.post(url, data=data)

# # Print the response content
# print(response.text)

### 4A: Modify your Flask endpoint to accept JSON data

In [15]:
from flask import Flask, request, jsonify

app = Flask(__name__)

@app.route('/submit', methods=['GET', 'POST'])
def submit():
    if request.method == 'POST':
        data = request.get_json()
        message = data.get('message', 'No message')
        return jsonify({'response': f'You submitted: {message}'})
    else:
        return '''
            <form method="post" action="/submit">
                <input type="text" name="message" placeholder="Enter a message" />
                <input type="submit" />
            </form>
        '''

if __name__ == '__main__':
    app.run(debug=False)


 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [30/Oct/2024 17:38:50] "POST /submit HTTP/1.1" 200 -


In [ ]:
# import requests

# url = 'http://127.0.0.1:5000/submit'

# json_data = {
#     'message': 'Hello, Flask with JSON!'
# }

# response = requests.post(url, json=json_data)

# print(response.json())


# 5. Advanced APP Route using HTML templates and CSS rendering

In [ ]:
from flask import Flask, request, render_template
import math

app = Flask(__name__)

@app.route('/')
def home():
    return render_template('home.html')

@app.route('/sqrt', methods=['GET','POST'])
def compute_sqrt():
    if request.method == 'POST':
        try:
            num1 = request.form.get('num1', type=float)
            num2 = request.form.get('num2', type=float)

            if num1 is None or num2 is None:
                error = 'Both Number 1 and Number 2 are required.'
                return render_template('form.html', error=error)

            if num2 == 0:
                error = 'Division by zero is not allowed.'
                return render_template('form.html', error=error)

            division_result = num1 / num2

            if division_result < 0:
                error = 'Cannot compute square root of a negative number.'
                return render_template('form.html', error=error)

            sqrt_result = math.sqrt(division_result)

            return render_template(
                'result.html',
                division_result=division_result,
                sqrt_result=sqrt_result
            )

        except (ValueError, TypeError):
            error = 'Invalid input. Please provide numeric values for Number 1 and Number 2.'
            return render_template('form.html', error=error)
    else:
        return render_template('form.html')

if __name__ == '__main__':
    app.run(debug=False)

# http://127.0.0.1:5000


 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [17/Dec/2025 02:57:32] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [17/Dec/2025 02:57:32] "GET /static/styles.css HTTP/1.1" 200 -


- app.py: The main Flask application script.
- static/: Contains static files like CSS, JavaScript, images, etc. The styles.css file is placed here.

- templates/: Contains HTML templates rendered by Flask.

- form.html is the template for the form page.
- result.html is the template for displaying the computation result.

## **Understanding the `304 Not Modified` Status Code**

### **What is `304 Not Modified`?**

- **Definition:**
  - The `304 Not Modified` is an HTTP status code indicating that the requested resource has not been modified since the last time it was requested.
- **Purpose:**
  - It tells the client (usually a web browser) that it can use the cached version of the resource, reducing the need to download it again.


### **Benefits of `304 Not Modified`:**

- **Reduced Server Load:**
  - Less data is sent over the network.
- **Faster Load Times:**
  - The browser can render the page more quickly using cached resources.
- **Optimized Bandwidth Usage:**
  - Particularly beneficial for users with limited or metered internet connections.

